powder_reduction.py

```diff
@@ -1,37 +1,142 @@
+#!/usr/bin/env python
+# coding: utf-8
+import numpy as np
 import scipp as sc
 from scipp import Dim
+from scipp.neutron.diffraction import load_calibration
+import smooth_data
 
+LAMBDA_RANGE = (0.7, 10.35)
 
-def powder_reduction(data, calibration=None):
+
+def powder_reduction(sample='sample.nxs', calibration=None):
     """
-    Takes a dataset with tof data and detector positions and converts to
-    d-spacing. Can use calibration data which needs to contain the fields
-    difc and tzero.
+    Simple WISH reduction workflow
+
+    Corrections applied:
+    - Normalization by monitors
+    - Conversion considering calibration
+    - Masking and grouping detectors into panels
 
     Parameters
     ----------
-        data: scipp Dataset
-            The input data to be converted
+    sample: Nexus event file
 
-    Keyword parameters
-    ------------------
-        calibration: scipp Dataset
-            Dataset containing difc and tzero for all detectors
+    calibration: .cal file following Mantid's standards
+        The columns correspond to detectors' IDs, offset, selection of detectors
+        and groups
+
+    Returns
+    -------
+    Scipp dataset containing reduced data in d-spacing
 
     """
 
+    # Load data
+    sample = sc.neutron.load(sample,
+                             load_pulse_times=False,
+                             mantid_args={'LoadMonitors': True})
+
+    # Load calibration
+    if calibration is not None:
+        input_load_cal = {"InstrumentName": "WISH"}
+        cal = load_calibration(calibration, mantid_args=input_load_cal)
+        # Merge table with detector->spectrum mapping from sample
+        # (implicitly checking that detectors between sample and calibration are the same)
+        # cal = sc.merge(cal, sample.labels['detector_info'].value)
+        # Compute spectrum mask from detector mask
+        mask = sc.groupby(cal['mask'], group='spectrum', combine=Dim.Spectrum).any(Dim.Detector)
```

Comment:
I thought there is no `spectrum` column in the table. How can the `groupby` work without the `merge`?

Reply:
It is my mistake: while preparing the file, line 26 was commented out. It should not be